utils/ai/index.ts

```
@@ -0,0 +1,234 @@
+  }
+
+  const { appId, userId } = userData;
+
+  // Fetch conversation history
+  const {
+    messages: messageHistory,
+    thoughts: thoughtHistory,
+    honchoMessages: honchoHistory,
+    pdfMessages: pdfHistory,
+    summaries: summaryHistory,
+    collectionId: existingCollectionId,
+  } = await fetchConversationHistory(appId, userId, conversationId);
+
+  // Generate thought
+  const thoughtPrompt = buildThoughtPrompt(
+    messageHistory,
+    thoughtHistory,
+    honchoHistory,
+    pdfHistory,
+    message,
+    Boolean(fileContent || existingCollectionId)
+  );
+  const { textStream: thoughtStream } = streamText({
+    messages: thoughtPrompt,
+    metadata: {
+      sessionId: conversationId,
+      userId,
+      type: 'thought',
+    },
+  });
+
+  let thought = '';
+  for await (const chunk of thoughtStream) {
+    thought += chunk;
+    yield formatStreamChunk({
+      type: 'thought',
+      text: chunk,
+    });
+  }
+
+  const [honchoContent, { pdfContent, collectionId }] = await Promise.all([
+    // HONCHO STUFF
+    (async () => {
+      const honchoQuery = extractTagContent(thought, 'honcho');
+      const { content: honchoContent } = await honcho.apps.users.sessions.chat(
+        appId,
+        userId,
+        conversationId,
+        { queries: honchoQuery }
+      );
+      return honchoContent;
+    })(),
+    // PDF STUFF
+    (async () => {
+      // Get PDF response if needed
+      let pdfContent = '';
+      let collectionId: string | undefined;
+      const fileContentArray = await fileContent;
+      if (fileContentArray || existingCollectionId) {
+        // If we have a new file, create a collection and add documents
+        if (fileContentArray) {
+          const collection = await honcho.apps.users.collections.create(
+            appId,
+            userId,
+            {
+              name: `PDF Collection - ${conversationId}`,
+            }
+          );
+          collectionId = collection.id;
```

This will throw an error that is not shown to the user if you try to submit an additional PDF to the chat, we should use get_or_create for getting the collection and use the collection metadata for tracking the total mb file size uploaded to the collection. 

I think restricting it to 5mb probably makes sense. 

Alternatively we could report an error to the end user that says they can only upload one file at a time

---

handled at upload time. the pdf sizes compress a ton once they've been ocr'd down to text so unlikely that cumulative file size will ever reach 5mb, but theres another check there.